### PR TITLE
ExpirationMap.

### PR DESCRIPTION
**Pull request by brian-js**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/routing/732)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @benjaminbollen

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/utilities/expiration_map.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+extern crate time;
```
Comment (benjaminbollen) [inline src/utilities/expiration_map.rs]: is this not already included in lib.rs?

File: src/utilities/expiration_map.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        match self.map.remove(key) {
+            Some((value, _)) => Some(value),
+            None => None,
+        }
+    }
+
+    /// Retrieves a value for the given key if present in the map.
+    pub fn get(&mut self, key: &K) -> Option<&V> {
+        match self.map.get(key) {
+            Some(&(ref value, _)) => Some(value),
```
Comment (benjaminbollen) [inline src/utilities/expiration_map.rs]: should the expiration not be validated on get and ignored if expired?  similar for contains_key.  On remove we can silently remove and has_expired is then not needed.  Sounds good? :)

File: src/utilities/expiration_map.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            assert!(expiration_map.get(&i).is_none());
+        }
+
+        assert!(expiration_map.contains_key(&11));
+        assert_eq!(expiration_map.get(&11).unwrap(), &11);
+    }
+
+    #[test]
+    fn insert_same_key_with_different_value() {
+        let duration = ::time::Duration::milliseconds(500);
```
Comment (benjaminbollen) [inline src/utilities/expiration_map.rs]: can we reduce this to 10 ms?

File: src/utilities/expiration_map.rs (hunk 2 above)
Comment (brian-js) [inline src/utilities/expiration_map.rs]: Already done.

File: src/utilities/expiration_map.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            None => false
+        }
+    }
+
+    /// Recover expired key-value pairs removing any such from the map.
+    pub fn remove_expired(&mut self) -> Option<Vec<(K,V)>> {
+        let mut expired = Vec::new();
+
+        for (key, &(ref value, time)) in self.map.iter() {
+            if time + self.time_to_live < ::time::SteadyTime::now() {
```
Comment (benjaminbollen) [inline src/utilities/expiration_map.rs]: probably better to only call ::time::SteadyTime::now() only once, but good for now

**Comment by benjaminbollen** [comment]
looks good to me

**Pull request merged**